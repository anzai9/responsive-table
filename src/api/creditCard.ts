import type { Paths } from "@/lib/typeUtils";

export const HOST = "https://my-json-server.typicode.com";
export const CREDIT_CARD_RESOURCE_PATH =
  "/liuderchi/json-server-v1/creditCards";

export type CreditCard = {
  title: string;
  image: string;
  score: string;
  network: string;
  isBusinessCard: number;
  reward?: {
    signUpBonus?: string;
    pointsCurrency?: string;
  };
};

export type CreditCardPath = NonNullable<Paths<CreditCard>>;

export const getCreditCards = async (): Promise<CreditCard[] | undefined> => {
  try {
    const response = await fetch(`${HOST}${CREDIT_CARD_RESOURCE_PATH}`);
    const data = await response.json();
    return data?.creditCards ?? [];
  } catch (err: unknown) {
    if (err instanceof Error) {
      throw new Error(err.message);
    }
  }
};
