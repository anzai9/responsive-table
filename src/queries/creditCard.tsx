import { useQuery } from "@tanstack/react-query";

import { getCreditCards } from "@/api/creditCard";

export const useCreditCards = () => {
  return useQuery({
    queryKey: ["creditCards"],
    queryFn: getCreditCards,
  });
};
