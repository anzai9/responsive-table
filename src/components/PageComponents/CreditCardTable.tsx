import { cn } from "@/lib/styleUtils";
import { useCreditCards } from "@/queries/creditCard";
import type { CreditCard, CreditCardPath } from "@/api/creditCard";

type Column = {
  key: CreditCardPath[];
  header: () => React.ReactNode;
  cell: (values: Record<string, string>) => React.ReactNode;
  enableFreeze?: boolean;
};

const DefaultHeader = ({
  title,
  className,
}: {
  title: string;
  className?: string;
}) => {
  return (
    <div className={cn("p-2 w-36 sm:w-48 box-border", className)}>{title}</div>
  );
};

const DefaultCell = ({ children }: { children: React.ReactNode }) => {
  return <div className="p-2 w-36 sm:w-48 box-border">{children}</div>;
};

const creditCardColumns: Column[] = [
  {
    key: ["title", "image"],
    header: () => (
      <DefaultHeader title="Credit Card" className="w-40 sm:w-60" />
    ),
    cell: ({ title, image }) => {
      return (
        <div className="p-2 w-40 sm:w-60 box-border">
          <img src={image} alt={title} className="w-48 h-auto aspect-4/3" />
          <p className="mt-2">{title}</p>
        </div>
      );
    },
    enableFreeze: true,
  },
  {
    key: ["score"],
    header: () => <DefaultHeader title="Recommend Credit Score" />,
    cell: ({ score }) => {
      return <DefaultCell>{score}</DefaultCell>;
    },
  },
  {
    key: ["isBusinessCard"],
    header: () => <DefaultHeader title="Is business credit card" />,
    cell: ({ isBusinessCard }) => {
      return <DefaultCell>{isBusinessCard ? "Yes" : "No"}</DefaultCell>;
    },
  },
  {
    key: ["reward.signUpBonus"],
    header: () => <DefaultHeader title="Welcome Offer" />,
    cell: ({ signUpBonus }) => {
      return <DefaultCell>{signUpBonus}</DefaultCell>;
    },
  },
  {
    key: ["reward.pointsCurrency"],
    header: () => <DefaultHeader title="Reward Currency" />,
    cell: ({ pointsCurrency }) => {
      return <DefaultCell>{pointsCurrency}</DefaultCell>;
    },
  },
  {
    key: ["network"],
    header: () => <DefaultHeader title="Merchant System" />,
    cell: ({ network }) => {
      return <DefaultCell>{network}</DefaultCell>;
    },
  },
];

const getCellValues = (creditCard: CreditCard, keys: CreditCardPath[]) => {
  return keys.reduce((acc, key) => {
    const chainingKeys = key.split(".");

    if (chainingKeys.length > 1) {
      let val: (CreditCard & CreditCard["reward"]) | string = creditCard;

      for (const k of chainingKeys) {
        if (typeof val === "object" && k in val) {
          val = val[k as "pointsCurrency" | "signUpBonus"] ?? "N/A";
        } else {
          val = "N/A";
          break;
        }
      }

      return { ...acc, [chainingKeys[chainingKeys.length - 1] as string]: val };
    }

    const val =
      creditCard[
        key as Exclude<
          CreditCardPath,
          "reward.pointsCurrency" | "reward.signUpBonus"
        >
      ];

    return { ...acc, [key]: val };
  }, {});
};

const sticky =
  "sticky left-0 border-2 border-l-0 after:content-[''] after:absolute after:right-[-2px] after:top-0 after:bottom-0 after:border-l-2 after:border-zinc-300 shadow-md shadow-zinc-300/50";

const CreditCardTableBody = () => {
  const { data, error, status } = useCreditCards();

  if (status === "pending") {
    return (
      <tbody>
        <tr>
          <td
            colSpan={creditCardColumns.length}
            className="h-12 text-lg text-center border-t-2 border-solid text-bold border-zinc-300"
          >
            Loading...
          </td>
        </tr>
      </tbody>
    );
  }

  return (
    <tbody>
      {(data?.length ?? 0) > 0 || error ? (
        data?.map((creditCard, I) => {
          return (
            <tr key={creditCard.title} className="group">
              {creditCardColumns.map((column) => {
                const key = column.key.join(".") + I;
                const vals = getCellValues(creditCard, column.key);
                return (
                  <td
                    key={key}
                    className={cn(
                      "h-auto text-left bg-white border-solid border-y-2 border-zinc-300 group-last:border-b-0",
                      {
                        [sticky]: column.enableFreeze,
                      },
                    )}
                  >
                    {column.cell({ ...vals })}
                  </td>
                );
              })}
            </tr>
          );
        })
      ) : (
        <tr>
          <td
            colSpan={creditCardColumns.length}
            className="h-12 text-lg text-center border-t-2 border-solid text-bold border-zinc-300"
          >
            No data
          </td>
        </tr>
      )}
    </tbody>
  );
};

export const CreditCardTable = () => {
  return (
    <div
      className="overflow-auto relative my-4 w-full border-2 border-solid border-zinc-300 box-border"
      data-testid="table-container"
    >
      <table className="relative w-full text-sm border-collapse">
        <thead>
          <tr className="group">
            {creditCardColumns.map((column) => {
              const key = column.key.join(".") + "header";
              return (
                <th
                  key={key}
                  className={cn(
                    "group-first:border-t-0 z-10 px-2 h-8 font-medium text-left align-middle bg-white border-t-0 border-solid border-y-2 text-stone-500 border-zinc-300",
                    {
                      [sticky]: column.enableFreeze,
                      "border-t-0": column.enableFreeze,
                    },
                  )}
                >
                  {column.header()}
                </th>
              );
            })}
          </tr>
        </thead>
        <CreditCardTableBody />
      </table>
    </div>
  );
};

export default CreditCardTable;
