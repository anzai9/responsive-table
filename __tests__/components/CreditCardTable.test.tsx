import { describe, expect, it } from "vitest";
import { QueryCache, QueryClient } from "@tanstack/react-query";
import { screen, waitFor, act } from "@testing-library/react";
import nock from "nock";

import CreditCardTable from "@/components/PageComponents/CreditCardTable";
import { CREDIT_CARD_RESOURCE_PATH, HOST } from "@/api/creditCard";
import { renderWithClient, sleep } from "../utils";

const mockData = [
  {
    title: "Capital One Venture X Rewards Credit Card Dynamic Page",
    image:
      "https://s.yimg.com/lo/api/res/1.2/BqLnLQeYGPL8v2po6QHKJg--/YXBwaWQ9ZWNfaG9yaXpvbnRhbDtzcz0x/https://cdn.prodstatic.com/shared/images/cards/401d0150-336c-11ec-b6bf-8d6df3a2e669.png.cf.jpg",
    score: "good",
    network: "Visa",
    isBusinessCard: 1,
    reward: {
      signUpBonus: "$200 statement credit after spending $1000 within 90 days",
    },
  },
  {
    title: "American Express Blue Business Cash Card _BP PDP",
    image:
      "https://s.yimg.com/lo/api/res/1.2/RItCSWKGFuSo4HnvDkYBOA--/YXBwaWQ9ZWNfaG9yaXpvbnRhbDtzcz0x/https://cdn.prodstatic.com/shared/images/cards/a10b5ad0-5d31-11ed-98ae-b771ec223ae8.webp.cf.jpg",
    score: "good",
    network: "Visa",
    isBusinessCard: 1,
    reward: {
      signUpBonus:
        "Earn 75,000 miles after spending $4,000 on purchases in the first 3 months",
      pointsCurrency: "Avios",
    },
  },
  {
    title: "Blue Cash Preferred Card from American Express _BP PDP",
    image:
      "https://s.yimg.com/lo/api/res/1.2/7_t3SAEFu3grjRRxpc8Rmg--/YXBwaWQ9ZWNfaG9yaXpvbnRhbDtzcz0x/https://cdn.prodstatic.com/shared/images/cards/d9b7b770-5d31-11ed-b47c-bfbb560b1ab3.webp.cf.jpg",
    score: "good",
    network: "Visa",
    isBusinessCard: 0,
  },
  {
    title: "British Airways Visa Signature Card",
    image:
      "https://s.yimg.com/lo/api/res/1.2/cQVw2qOdLyLnoPLTrxZbsg--/YXBwaWQ9ZWNfaG9yaXpvbnRhbDtzcz0x/https://cdn.prodstatic.com/shared/images/cards/2d8687e0-5d33-11ed-b47c-bfbb560b1ab3.webp.cf.jpg",
    score: "good",
    network: "Visa",
    isBusinessCard: 1,
  },
  {
    title: "OpenSky Secured VisaÂ® Credit Card",
    image:
      "https://s.yimg.com/lo/api/res/1.2/Ilj.pVqODGmB4HLqnG727A--/YXBwaWQ9ZWNfaG9yaXpvbnRhbDtzcz0x/https://cdn.prodstatic.com/shared/images/cards/7cb26200-bdbc-11ed-b5f1-9954c274724b.png.cf.jpg",
    score: "No credit check",
    network: "Mastercard",
    isBusinessCard: 0,
    reward: {
      signUpBonus: "$200 minimum security deposit",
    },
  },
  {
    title: "IHG One Rewards Premier Credit Card",
    image:
      "https://s.yimg.com/lo/api/res/1.2/6BFOcoiANtnAedDlaMU4bg--/YXBwaWQ9ZWNfaG9yaXpvbnRhbDtzcz0x/https://cdn.prodstatic.com/shared/images/cards/ce4c2920-1695-11ee-9d05-8969f2172cdd.webp.cf.jpg",
    score: "Excellent/Good",
    network: "Visa",
    isBusinessCard: 1,
    reward: {
      signUpBonus:
        "Earn 140,000 bonus points after spending $4,000 on purchases in the first 3 months",
      pointsCurrency: "Avios",
    },
  },
  {
    title: "Marriott Bonvoy Bevy American Express Card",
    image:
      "https://s.yimg.com/lo/api/res/1.2/Tf7r5_viMOBEpEJNvUtPpA--/YXBwaWQ9ZWNfaG9yaXpvbnRhbDtzcz0x/https://cdn.prodstatic.com/shared/images/cards/93fe6de0-39f1-11ed-b73c-91712e8727dd.png.cf.jpg",
    score: "good",
    network: "JCB",
    isBusinessCard: 1,
    reward: {
      signUpBonus:
        "Earn 2X points on all other eligible purchases (terms apply)",
    },
  },
];

describe("CreditCardTable", () => {
  const queryCache = new QueryCache();
  const queryClient = new QueryClient({
    queryCache,
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  });

  it.concurrent("should render expected order of table headers", async () => {
    nock(HOST)
      .get(CREDIT_CARD_RESOURCE_PATH)
      .reply(200, {
        creditCards: [mockData[0]],
      });

    renderWithClient(queryClient, <CreditCardTable />);

    expect(screen.getByRole("table")).toBeInTheDocument();

    const headers = screen.getAllByRole("columnheader");
    const expectedHeaderOrder = [
      "Credit Card",
      "Recommend Credit Score",
      "Is business credit card",
      "Welcome Offer",
      "Reward Currency",
      "Merchant System",
    ];
    headers.forEach((header, I) => {
      expect(header.textContent).toBe(expectedHeaderOrder[I]);
    });

    await waitFor(() =>
      expect(
        screen.getByText(
          "Capital One Venture X Rewards Credit Card Dynamic Page",
        ),
      ).toBeInTheDocument(),
    );

    expect(screen.getByText("good")).toBeInTheDocument();
    expect(screen.getByText("Yes")).toBeInTheDocument();
    expect(
      screen.getByText(
        "$200 statement credit after spending $1000 within 90 days",
      ),
    ).toBeInTheDocument();
    expect(screen.getByText("N/A")).toBeInTheDocument();
  });

  it("should render the No data while data is empty", async () => {
    nock(HOST).get(CREDIT_CARD_RESOURCE_PATH).reply(200, {
      creditCards: [],
    });

    renderWithClient(queryClient, <CreditCardTable />);

    expect(screen.getByRole("table")).toBeInTheDocument();

    await waitFor(() =>
      expect(screen.getByText("No data")).toBeInTheDocument(),
    );
  });

  it.concurrent("first column should be fixed", () => {
    nock(HOST)
      .get(CREDIT_CARD_RESOURCE_PATH)
      .reply(200, {
        creditCards: [mockData[0]],
      });
    renderWithClient(queryClient, <CreditCardTable />);
    const firstColumn = screen.getAllByRole("columnheader")[0];
    expect(firstColumn).toHaveClass("sticky");
    expect(firstColumn).toHaveClass("left-0");
  });

  it("should allow horizontal scrolling", async () => {
    nock(HOST)
      .get(CREDIT_CARD_RESOURCE_PATH)
      .reply(200, {
        creditCards: [mockData[0]],
      });
    renderWithClient(queryClient, <CreditCardTable />);
    const tableContainer = screen.getByTestId("table-container");
    expect(tableContainer).toHaveClass("overflow-auto");
  });

  it("should render data expectedly", async () => {
    nock(HOST).get(CREDIT_CARD_RESOURCE_PATH).reply(200, {
      creditCards: mockData,
    });
    const { container } = renderWithClient(queryClient, <CreditCardTable />);
    await act(async () => {
      await sleep(10);
    });
    expect(container).toMatchSnapshot();
  });
});
