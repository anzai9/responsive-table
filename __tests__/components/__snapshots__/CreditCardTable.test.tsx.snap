// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CreditCardTable > should render data expectedly 1`] = `
<div>
  <div
    class="overflow-auto relative my-4 w-full border-2 border-solid border-zinc-300 box-border"
    data-testid="table-container"
  >
    <table
      class="relative w-full text-sm border-collapse"
    >
      <thead>
        <tr
          class="group"
        >
          <th
            class="group-first:border-t-0 z-10 px-2 h-8 font-medium text-left align-middle bg-white border-solid border-y-2 text-stone-500 border-zinc-300 sticky left-0 border-2 border-l-0 after:content-[''] after:absolute after:right-[-2px] after:top-0 after:bottom-0 after:border-l-2 after:border-zinc-300 shadow-md shadow-zinc-300/50 border-t-0"
          >
            <div
              class="p-2 box-border w-40 sm:w-60"
            >
              Credit Card
            </div>
          </th>
          <th
            class="group-first:border-t-0 z-10 px-2 h-8 font-medium text-left align-middle bg-white border-solid border-y-2 text-stone-500 border-zinc-300"
          >
            <div
              class="p-2 w-36 sm:w-48 box-border"
            >
              Recommend Credit Score
            </div>
          </th>
          <th
            class="group-first:border-t-0 z-10 px-2 h-8 font-medium text-left align-middle bg-white border-solid border-y-2 text-stone-500 border-zinc-300"
          >
            <div
              class="p-2 w-36 sm:w-48 box-border"
            >
              Is business credit card
            </div>
          </th>
          <th
            class="group-first:border-t-0 z-10 px-2 h-8 font-medium text-left align-middle bg-white border-solid border-y-2 text-stone-500 border-zinc-300"
          >
            <div
              class="p-2 w-36 sm:w-48 box-border"
            >
              Welcome Offer
            </div>
          </th>
          <th
            class="group-first:border-t-0 z-10 px-2 h-8 font-medium text-left align-middle bg-white border-solid border-y-2 text-stone-500 border-zinc-300"
          >
            <div
              class="p-2 w-36 sm:w-48 box-border"
            >
              Reward Currency
            </div>
          </th>
          <th
            class="group-first:border-t-0 z-10 px-2 h-8 font-medium text-left align-middle bg-white border-solid border-y-2 text-stone-500 border-zinc-300"
          >
            <div
              class="p-2 w-36 sm:w-48 box-border"
            >
              Merchant System
            </div>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="group"
        >
          <td
            class="h-auto text-left bg-white border-solid border-y-2 border-zinc-300 group-last:border-b-0 sticky left-0 border-2 border-l-0 after:content-[''] after:absolute after:right-[-2px] after:top-0 after:bottom-0 after:border-l-2 after:border-zinc-300 shadow-md shadow-zinc-300/50"
          >
            <div
              class="p-2 w-40 sm:w-60 box-border"
            >
              <img
                alt="Capital One Venture X Rewards Credit Card Dynamic Page"
                class="w-48 h-auto aspect-4/3"
                src="https://s.yimg.com/lo/api/res/1.2/BqLnLQeYGPL8v2po6QHKJg--/YXBwaWQ9ZWNfaG9yaXpvbnRhbDtzcz0x/https://cdn.prodstatic.com/shared/images/cards/401d0150-336c-11ec-b6bf-8d6df3a2e669.png.cf.jpg"
              />
              <p
                class="mt-2"
              >
                Capital One Venture X Rewards Credit Card Dynamic Page
              </p>
            </div>
          </td>
          <td
            class="h-auto text-left bg-white border-solid border-y-2 border-zinc-300 group-last:border-b-0"
          >
            <div
              class="p-2 w-36 sm:w-48 box-border"
            >
              good
            </div>
          </td>
          <td
            class="h-auto text-left bg-white border-solid border-y-2 border-zinc-300 group-last:border-b-0"
          >
            <div
              class="p-2 w-36 sm:w-48 box-border"
            >
              Yes
            </div>
          </td>
          <td
            class="h-auto text-left bg-white border-solid border-y-2 border-zinc-300 group-last:border-b-0"
          >
            <div
              class="p-2 w-36 sm:w-48 box-border"
            >
              $200 statement credit after spending $1000 within 90 days
            </div>
          </td>
          <td
            class="h-auto text-left bg-white border-solid border-y-2 border-zinc-300 group-last:border-b-0"
          >
            <div
              class="p-2 w-36 sm:w-48 box-border"
            >
              N/A
            </div>
          </td>
          <td
            class="h-auto text-left bg-white border-solid border-y-2 border-zinc-300 group-last:border-b-0"
          >
            <div
              class="p-2 w-36 sm:w-48 box-border"
            >
              Visa
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
